Added average damage to tank-win-rate and win-rate

supinko opened the pull request:
The tank-win-rate and win-rate commands now add your average damage stats. If there are no recorded damage stats (as will be the case with everyone at first) then that part of the message is omitted.

supinko committed:
Added average damage to tank-win-rate and win-rate

CodeMan99 commented:
Nice. I'll try to review soon. Glad you updated the tests. :+1:

supinko commented:
Haha! Glad you actually have tests!
    On Saturday, March 23, 2019, 5:53:58 PM PDT, Cody A. Taylor <notifications@github.com> wrote:  
 
 
Nice. I'll try to review soon. Glad you updated the tests. 👍
 
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

CodeMan99 commented on lib/command/winRate.js:
```diff
@@ -68,7 +68,7 @@ function twrFn(msg, authorRecord, tankName) {
 			}
 
-			return this.wotblitz.tanks.stats(record.account_id, null, tankIds, null, ['tank_id', 'all.battles', 'all.wins'])
+			return this.wotblitz.tanks.stats(record.account_id, null, tankIds, null, ['tank_id', 'all.battles', 'all.wins', 'all.damage_dealt'])
```
I am pretty sure this makes you the first user of wotblitz. If you can remember, give me some idea how long it took you to identify what change to make here.

CodeMan99 commented on lib/command/winRate.js:
```diff
@@ -94,27 +95,44 @@ function twrFn(msg, authorRecord, tankName) {
 
 				percent = ((wins - record.wins) / (battles - record.battles)) * 100;
-				send += `\nOver those ${battles - record.battles} battles, you won ${percent.toFixed(2)}%!`;
+				avgDmg = ((damage - record.damage) / (battles - record.battles));
```
Unnecessary parens. Surprised I don't have an eslint rule for that.

```javascript
avgDmg = (damage - record.damage) / (battles - record.battles);
```

CodeMan99 commented on lib/command/winRate.js:
```diff
@@ -94,27 +95,44 @@ function twrFn(msg, authorRecord, tankName) {
+				if (avgDmg) {
+					send += ` with average damage of ${avgDmg.toFixed(2)}`;
+				}
+				send += `!`;
```
The punctuation was in the else block at line 118. Pick one pattern and use it in both places.

CodeMan99 commented on lib/command/winRate.js:
```diff
@@ -80,9 +80,10 @@ function twrFn(msg, authorRecord, tankName) {
 
 						return tankopedia.name + ' (' + tankopedia.nation + ', ' + tankopedia.tier + '): ' +
-							winRate.toFixed(2) + '%' + ' after ' + stat.all.battles + ' battles.';
+							winRate.toFixed(2) + '%, ' + avgDmg.toFixed(2) + ' dmg after ' + stat.all.battles + ' battles.';
```
Spell out "dmg" please.

CodeMan99 commented on test/lib.command.winRate.js:
```diff
@@ -440,22 +447,24 @@ test('command.winRate.winRate', t => {
-					'Over those 1 battles, you won 100.00%!'
+					'@biggirl22 [CL], You have won 9260 of 13933 battles. That is 66.46% victory! Your average damage is 1000.00.',
+					'Last time you asked was 1 battles ago, at 66.46% victory and 999.94 average damage dealt.',
+					'Over those 1 battles, you won 100.00% with average damage of 1836.00!'
```
Believe or not, users commonly run this command after only one battle. Please find a solution that is a bit more intelligent about decimal places.

For example
* A single battle should have a whole number.
* No trailing zeros beyond the natural result. So "1600.4" is written as `1600.4`, but "1600.404" is written as `1600.40`.
* Ignore floating point errors. So "1600.000000000001" is just `1600`.

CodeMan99 reviewed (changes requested):
Finally got around to looking at this. I'll want to merge this PR first since you included 8394861 in #15 as well.

supinko commented:
It was pretty easy; I went to the WG developer site, used their REST API tester to see what properties were available, and added it here.
    On Sunday, April 14, 2019, 8:41:49 PM PDT, Cody A. Taylor <notifications@github.com> wrote:  
 
 
@CodeMan99 commented on this pull request.

In lib/command/winRate.js:
> @@ -68,7 +68,7 @@ function twrFn(msg, authorRecord, tankName) {
 				});
 			}
 
-			return this.wotblitz.tanks.stats(record.account_id, null, tankIds, null, ['tank_id', 'all.battles', 'all.wins'])
+			return this.wotblitz.tanks.stats(record.account_id, null, tankIds, null, ['tank_id', 'all.battles', 'all.wins', 'all.damage_dealt'])

I am pretty sure this makes you the first user of wotblitz. If you can remember, give me some idea how long it took you to identify what change to make here.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

supinko commented:
Thanks for the review! I'll see about making your requested changes.
    On Sunday, April 14, 2019, 9:04:08 PM PDT, Cody A. Taylor <notifications@github.com> wrote:  
 
 
@CodeMan99 requested changes on this pull request.

Finally got around to looking at this. I'll want to merge this PR first since you included 8394861 in #15 as well.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

supinko commented:
Yeah, the number formatting was a real bugger. I'm new to Javascript so I just copied your use of Number.toFixed(). I'll give Intl.NumberFormat a shot.
    On Sunday, April 14, 2019, 9:01:23 PM PDT, Cody A. Taylor <notifications@github.com> wrote:  
 
 
@CodeMan99 commented on this pull request.

In test/lib.command.winRate.js:
>  		}).then(result => {
 			st.deepEqual(result, {
 				sentMsg: [
-					'@biggirl22 [CL], You have won 9260 of 13933 battles. That is 66.46% victory!',
-					'Last time you asked was 1 battles ago, at 66.46% victory.',
-					'Over those 1 battles, you won 100.00%!'
+					'@biggirl22 [CL], You have won 9260 of 13933 battles. That is 66.46% victory! Your average damage is 1000.00.',
+					'Last time you asked was 1 battles ago, at 66.46% victory and 999.94 average damage dealt.',
+					'Over those 1 battles, you won 100.00% with average damage of 1836.00!'

Believe or not, users commonly run this command after only one battle. Please find a solution that is a bit more intelligent about decimal places.

For example
   
   - A single battle should have a whole number.
   - No trailing zeros beyond the natural result. So "1600.4" is written as 1600.4, but "1600.404" is written as 1600.40.
   - Ignore floating point errors. So "1600.000000000001" is just 1600.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

supinko committed:
Addressed code review comments.

Added a number formatting function to make the non-percentage numbers in the win-rate command look better.

supinko committed:
Added a timestamp to the log messages so it's easier to debug problems.

supinko committed:
Moved the formatNumber function out to helpers.js and applied some new formatting to the maxXp command to make it more readable. If you disagree with the new formatting, I can revert it.

CodeMan99 commented on bin/bot.js:
```diff
@@ -73,7 +73,8 @@ client.on('message', message => {
 	const userId = message.author.id;
-	const id = message.id + ', ' + userId;
+	const userName = message.author.username;
+	const id = 'msgId: ' + message.id + ', userId: ' + userId + ', userName: ' + userName;
```
Do *not* log the username. The `userId` is well more than enough to identify a person. There are three reasons for this.

1. A username may contain non-printable characters, thus making the logs *less* searchable.
2. I don't want to dig into the developer agreement to find out if storing the a discord username is allowable.
3. The username can be changed at any time, while the id value will (probably) never change.

Last, the whole point of `id` to quickly identify a message in the logs by being highly searchable. Discord has a [developer mode](https://support.discordapp.com/hc/en-us/articles/206346498-Where-can-I-find-my-User-Server-Message-ID-) that allows you to easily copy these id values.

supinko replied:
I guess it depends on how you want to use the log file. If you want to throw it in Splunk or some other search/analysis engine, then just having the ID is fine, but I'd argue that you can't really do that effectively w/o the timestamp (added below). Also, if you just want to see, "what did this person do?" then just the ID is fine. However, if you want to see, "who did what and when?" then having the user name, even if it changes occasionally or has some non-printable characters, is essential.

Notice I didn't remove the ID; I just added the user name.

Regarding whether the user name is allowed to be logged, you're already logging the user ID, which is personally identifiable information (PII), so the user name is no worse. 

Having said all this, if you really object to the user name, I'll remove it.

CodeMan99 replied:
I do object. I do not feel it adds any transparency.

More important would be logging guild and channel information. However, I do not want that either because two more id values would effectively increase the byte growth of the log file by about 30%. Additionally, you can look up this values with the back end REPL.

CodeMan99 commented on lib/helpers.js:
```diff
@@ -84,3 +87,47 @@ function messageToString(message) {
+ */
+function log(message) {
+	console.log(new Date().toLocaleString() + ' ' + messageToString(message));
+}
```
Not sure the date prefix is actually useful. Mind giving your reason for adding it?

Additionally, logging is intentionally only done in _bin/bot.js_ so if there is decent reason, just prefix the `id` variable.

supinko replied:
I had a case where I noticed a problem days after it happened and I wasn't able to correlate the log messages to the original request. Adding the timestamp let me do that.

I can't just prefix the `id` variable because not all log messages use it (errors, or the startup message, for example)

supinko replied:
BTW, I moved the `log` function to helpers because I was using it in a command file in another branch that I'm working on. What's the reason for only logging in bin/bot.js?

supinko replied:
Never mind. I'll live w/o the log function.

CodeMan99 replied:
> What's the reason for only logging in bin/bot.js?

So logging is centralized and commands are more sandboxed.

supinko committed:
Removed user name from log message. Moved log and error functions out of helpers.js and into bin/bot.js.

CodeMan99 commented on bin/bot.js:
```diff
@@ -205,8 +205,28 @@ Promise.all([
+ *
+ * @param {*} message A message of any type
+ */
+function log(message) {
```
My request was to discuss these helpers. I did not intend for you to move them.

CodeMan99 commented:
Do I need to provide more direction here? I'd really like to release this feature!
